File: Level_1/크레인 인형뽑기.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>

using namespace std;

int solution(vector<vector<int>> board, vector<int> moves) {
    
    stack<int> s;
    int answer = 0;
    
    for(int i = 0; i < moves.size(); ++i) {
        int level = 0, doll = -1;
        bool check = false;
        while(board[level][moves[i]-1] == 0) {
            if(level == board.size()-1) {
                check = true;
                break;
            }
            ++level;
        }
        // 해당 위치에 인형이 없는 경우
        if(check) continue;
        
        doll = board[level][moves[i]-1];
        board[level][moves[i]-1] = 0;
        if(!s.empty() && s.top() == doll) {
            answer += 2;
            s.pop();
        }
        else s.push(doll);
    }
    return answer;
}

File: 2019 카카오 개발자 겨울 인턴십/크레인 인형뽑기.java
import java.util.*;

class Solution {
    public int solution(int[][] board, int[] moves) {
        int n = board.length, answer = 0;
        Stack<Integer> mystack = new Stack<>();
        for (int move : moves) {
            int pos = move - 1;
            for (int r = 0; r < n; ++r) {
                if (board[r][pos] != 0) {
                    if (mystack.empty()) {
                        mystack.push(board[r][pos]);
                    }
                    else {
                        int top = mystack.peek();
                        if (board[r][pos] == top) {
                            mystack.pop();
                            answer += 2;
                        }
                        else mystack.push(board[r][pos]);
                    }
                    board[r][pos] = 0;
                    break;
                }
            }
        }
        return answer;
    }
}
